addons/website/static/src/snippets/s_website_form/000.js

```diff
@@ -719,4 +773,94 @@ import { sprintf } from "@web/core/utils/strings";
+         * delete them) and manages the files.
+         *
+         * @private
+         * @param {Event} ev
+         */
+        _onFileChange(ev) {
+            const fileInputEl = ev.currentTarget;
+            const fieldEl = fileInputEl.closest('.s_website_form_field');
+            const uploadedFiles = fileInputEl.files;
+            const addFilesButtonEl = fieldEl.querySelector('.o_add_files_button');
+
+            // The zone where the file blocks are displayed.
+            let filesZoneEl = fieldEl.querySelector('.o_files_zone');
+            // Update the button text content.
+            if (!addFilesButtonEl) {
+                this._createAddFilesButton(fileInputEl);
+            }
+
+            // Create a list to keep track of the files.
+            if (!fileInputEl.fileList) {
+                fileInputEl.fileList = new DataTransfer();
+            }
+
+            // If only one file can be uploaded, delete the previous file.
+            if (!fileInputEl.getAttribute('multiple') && uploadedFiles.length > 0) {
+                fileInputEl.fileList = new DataTransfer();
+                const fileBlockEl = fieldEl.querySelector('.o_file_block');
+                if (fileBlockEl) {
+                    fileBlockEl.remove();
+                }
+            }
+
+            // Add the uploaded files if they are not already there.
+            for (const newFile of uploadedFiles) {
+                if (![...fileInputEl.fileList.files].some(file => newFile.name === file.name && newFile.size === file.size && newFile.type === file.type)) {
+                    fileInputEl.fileList.items.add(newFile);
+                    const fileDetails = {name: newFile.name, size: newFile.size, type: newFile.type};
+                    this._createFileBlock(fileDetails, filesZoneEl);
+                }
+            }
+            // Update the input files.
+            fileInputEl.files = fileInputEl.fileList.files;
+        },
+        /**
+         * Called when a file is deleted by clicking on the cross on the block
+         * describing it.
+         *
+         * @private
+         * @param {Event} ev
+         */
+        _onFileDelete(ev) {
+            const fileBlockEl = ev.target.closest('.o_file_block');
+            const fieldEl = fileBlockEl.closest('.s_website_form_field');
+            const fileInputEl = fieldEl.querySelector('input[type="file"]');
+            const fileDetails = fileBlockEl.fileDetails;
+            const addFilesButtonEl = fieldEl.querySelector('.o_add_files_button');
+
+            // Create a new file list containing the remaining files.
+            const newFileList = new DataTransfer();
+            for (const file of Object.values(fileInputEl.fileList.files)) {
+                if (file.name !== fileDetails.name || file.size !== fileDetails.size || file.type !== fileDetails.type) {
+                    newFileList.items.add(file);
+                }
+            }
```

Comment:
Can't you use `filter()` here ?

Reply:
No because `files` is an object so I can't use `filter` on it. I could convert it to an array before (with `[...fileInputEl.fileList.files]`) and then use filter but the result would be an array and I would still need to browse it to add the files in `newFilelist` one by one :)
So there would be two iterations (the one in filter and then the addition of files) while only one is needed here :) (it doesn't improve the complexity since the two iterations are one after the other but I think one iteration is better than two)